src/Features/Core/Portable/ConvertForEachToFor/AbstractConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,474 @@
+        {
+            var (start, end) = GetForEachBody(foreachStatement);
+            if (start == null || end == null)
+            {
+                // empty body. this can happen in VB
+                return false;
+            }
+
+            var dataFlow = semanticModel.AnalyzeDataFlow(start, end);
+
+            if (!dataFlow.Succeeded)
+            {
+                // if we can't get good analysis, assume it is written
+                return true;
+            }
+
+            return dataFlow.WrittenInside.Contains(foreachVariable);
+        }
+
+        private async Task<Document> ConvertForeachToForAsync(
+            Document document,
+            ForEachInfo foreachInfo,
+            CancellationToken cancellationToken)
+        {
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var editor = new SyntaxEditor(model.SyntaxTree.GetRoot(cancellationToken), workspace);
+
+            ConvertToForStatement(model, foreachInfo, editor, cancellationToken);
+
+            var newRoot = editor.GetChangedRoot();
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected class ForEachInfo
+        {
+            public ForEachInfo(
+                ISemanticFactsService semanticFacts, OptionSet options,
+                string collectionNameSuggestion, string countName,
+                ITypeSymbol explicitCastInterface, ITypeSymbol forEachElementType,
+                bool requireCollectionStatement, TForEachStatement forEachStatement)
+            {
+                SemanticFacts = semanticFacts;
+                Options = options;
+
+                CollectionNameSuggestion = collectionNameSuggestion;
+                CountName = countName;
+
+                // order of setting properties is important here
+                ExplicitCastInterface = explicitCastInterface;
+                ForEachElementType = forEachElementType;
+
+                RequireCollectionStatement = requireCollectionStatement || RequireExplicitCastInterface;
+
+                ForEachStatement = forEachStatement;
+            }
+
+            public bool RequireExplicitCastInterface => ExplicitCastInterface != null;
+
+            public ISemanticFactsService SemanticFacts { get; }
+            public OptionSet Options { get; }
+
+            public string CollectionNameSuggestion { get; }
+            public string CountName { get; }
+            public ITypeSymbol ExplicitCastInterface { get; }
+            public ITypeSymbol ForEachElementType { get; }
+            public bool RequireCollectionStatement { get; }
+            public TForEachStatement ForEachStatement { get; }
+        }
+
+        private class ForEachToForCodeAction : CodeAction.DocumentChangeAction
```

Comment:
note: IDE pattern is to name these MyCodeAction in almost all cases.  That said, that name is pretty terrible.  So if you want to keep the name like this, i'm ok with that.